06_gpu_and_ml/stable_lm/main.py

```diff
@@ -0,0 +1,322 @@
+import base64
+import os
+import time
+from pathlib import Path
+from typing import Any, Dict, Generator, List, Optional, Union
+
+from pydantic import BaseModel
+from typing_extensions import Annotated, Literal
+
+import modal
+
+with open("requirements.txt", "r") as f:
+    requirements = base64.b64encode(f.read().encode("utf-8")).decode("utf-8")
+
+
+def build_models():
+    import torch
+    from huggingface_hub import snapshot_download
+    from transformers import AutoModelForCausalLM, AutoTokenizer
+
+    model_path = snapshot_download(
+        "stabilityai/stablelm-tuned-alpha-7b",
+        ignore_patterns=["*.md"],
+    )
+    m = AutoModelForCausalLM.from_pretrained(
+        model_path,
+        torch_dtype=torch.float16,
+        device_map="auto",
+        local_files_only=True,
+    )
+    m.save_pretrained(
+        model_path, safe_serialization=True, max_shard_size="24GB"
+    )
+    tok = AutoTokenizer.from_pretrained(model_path)
+    tok.save_pretrained(model_path)
+    [p.unlink() for p in Path(model_path).rglob("*.bin")]  # type: ignore
+
+
+image = (
+    modal.Image.conda()
+    .apt_install("git", "software-properties-common", "wget")
+    .conda_install(
+        "cudatoolkit-dev=11.7",
+        "pytorch-cuda=11.7",
+        "rust=1.69.0",
+        channels=["nvidia", "pytorch", "conda-forge"],
+    )
+    .env(
+        {
+            "HF_HOME": "/root",
+            "HF_HUB_ENABLE_HF_TRANSFER": "1",
+            "SAFETENSORS_FAST_GPU": "1",
+            "BITSANDBYTES_NOWELCOME": "1",
+            "PIP_DISABLE_PIP_VERSION_CHECK": "1",
+            "PIP_NO_CACHE_DIR": "1",
+        }
+    )
+    .run_commands(
+        f"echo '{requirements}' | base64 --decode > /root/requirements.txt",
+        "pip install -r /root/requirements.txt",
+        gpu="any",
+    )
+    .run_function(
+        build_models,
+        gpu=None,
+        timeout=3600,
+    )
+)
+
+stub = modal.Stub(
+    name="example-stability-lm",
+    image=image,
+    secrets=[modal.Secret({"REPO_ID": "stabilityai/stablelm-tuned-alpha-7b"})],
```

Comment:
What's this env var used for? I see that this ID string appears in a number of other places, so wonder if this one is redundant.

Reply:
Good point it's redundant. I was trying to solve the original issue where I was re-writing this in a few places by putting in main stub but didn't end up using :) Ideally we can abstract a bit more to handle passing in various models and sizes as they are released.